automata/base/automaton.py

```diff
@@ -22,39 +34,40 @@ def __init__(self, **kwargs):
                 object.__setattr__(self, attr_name, attr_value)
         self.__post_init__()
 
-    def __post_init__(self):
+    def __post_init__(self) -> None:
         if global_config.should_validate_automata:
             self.validate()
 
     @abc.abstractmethod
-    def validate(self):
+    def validate(self) -> bool:
         """Return True if this automaton is internally consistent."""
         raise NotImplementedError
 
-    def __setattr__(self, name, value):
+    def __setattr__(self, name: str, value: Any) -> NoReturn:
         """Set custom setattr to make class immutable."""
         raise AttributeError(f'This {type(self).__name__} is immutable')
 
-    def __delattr__(self, name):
+    def __delattr__(self, name: str) -> None:
         """Set custom delattr to make class immutable."""
         raise AttributeError(f'This {type(self).__name__} is immutable')
 
-    def __getstate__(self):
+    def __getstate__(self) -> Any:
         """Return the object's state, described by its input parameters"""
         return self.input_parameters
 
-    def __setstate__(self, d):
+    def __setstate__(self, d: Dict[str, Any]) -> None:
         """Restore the object state from its input parameters"""
         # Notice that the default __setstate__ method won't work
         #   because __setattr__ is disabled due to immutability
-        self.__init__(**d)
+        self.__init__(**d) # type: ignore
```

Comment:
@eliotwrobson What is the purpose of the ignore for this particular line?

Reply:
This is the error that gets raised:

![image](https://user-images.githubusercontent.com/1345068/229000949-e1144e9d-1c2a-48ef-a663-941fae8cf7f8.png)

Not quite sure what this particular function is doing, so it may make sense to do a minor refactor to avoid this unsound access.

Reply:
@eliotwrobson Defining the `__getstate__` and `__setstate__` methods enable the automaton objects to be picked—this ability was added back in #125, and is one I'd like to keep (especially since it was a requested feature).

After reviewing that PR, it seems the pickling capability only really applies to mutable automaton instances anyway (i.e. with `global_config.allow_mutable_automata = true`). So I don't think we can use `self.__class__(...)` like the `copy` method does, since that creates a new instance.

Reply:
That makes sense, having the ignore here isn't a huge deal and I agree that pickling support is a feature worth keeping.